### ISSUE TITLE
Eliminate duplicate `@Nullable` annotations

### ISSUE THREAD

**Issue by donalmurtagh** [jspecify/jspecify#713]
I'm using [NullAway](https://github.com/uber/NullAway) to verify my nullness definitions. By default, everything is assumed to be not-null, so it's necessary to add `@Nullable` to method arguments and return types than can be null, e.g.

```java
public class ResourceConverters {
    public static @Nullable Resource toResource(@Nullable MultipartFile file) {
        return file != null ? file.getResource() : null;
    }

    public static @Nullable NamedByteArrayResource toNamedByteArrayResource(@Nullable MultipartFile file) {
        return file != null ? new NamedByteArrayResource(file.getOriginalFilename(), file.getBytes()) : null;
    }
}
```

Is there a way to avoid duplicating the `@Nullable` annotation for every parameter and return type? 

When `@NullMarked` is used at the class-level it effectively applies a default of **not-null** to all method parameters and return values in that class. What I'm looking for is the opposite, i.e. a way to apply a default of nullable at the class level.

I'm aware of `@NullUnmarked`, but this seems to apply a default of "nullness unknown", but what I want is a default of "known to be nullable".

**Comment by agentgt** [comment]
NullUnmarked is your best bet. You are right that is technically not the same but I think it effectively works out to be mostly the negation of NullMarked.

BTW If this is new code I highly recommend you do not follow that pattern of nullable in and nullable out especially for public methods. Technically that code is `@PolyNull`.

See Apache Commons used to do that all the time and before Guava came out it was pretty detrimental (Guava does not do nullable in and nullable out everywhere like commons). This is one of the few strong opinions I have but is based on lots of experience. 

See what happens is because nobody wants to deal with the `null` it just gets passed along multiple times and when you do get a NPE it goes from normally being an easy bug to fix to where in the hell in the stack did it get set to `null`.

**Comment by donalmurtagh** [comment]
> BTW If this is new code I highly recommend you do not follow that pattern of nullable in and nullable out especially for public methods.

@agentgt Presumably you think I should return `Optional` instead of null?

My weakly-held view on the use of optional is that it should be used (instead of a nullable value), if the returned value is ever dereferenced. On the other hand, if the empty optional is always unwrapped to a null value, I don't see much point in using an optional. Here's an example

## With Optional
```java
void Optional<Object> getOptionalValue() {
  return Optional.ofNullable(somethingThatMightBeNull);
}

// somewhere in the codebase
var val1 = getOptionalValue().orElse(null);

// somewhere else in the codebase
var val2 = getOptionalValue().orElse(null);
```

## Without Optional
```java
void Object getValue() {
  return somethingThatMightBeNull;
}

// somewhere in the codebase
var val1 = getValue();

// somewhere else in the codebase
var val2 = getValue();
```

If I know that:
 - Everywhere `getOptionalValue` returns an empty optional, it is unwrapped to a null
 - I expect this to remain true forever

Then I don't see much value in using the first example instead of the second.

**Comment by donalmurtagh** [comment]
> NullUnmarked is your best bet. You are right that is technically not the same but I think it effectively works out to be mostly the negation of NullMarked.

@agentgt Based on my testing, using a class-level `@NullUnmarked` is not the same as annotating each parameter/return value with `@Nullable`. Specifically, when I use this I correctly get warned by NullAway

```java
public class ResourceConverters {
    public static @Nullable Resource toResource(@Nullable MultipartFile file) {
        System.out.println("file = " + file.toString()); // potential NPE
        return file != null ? file.getResource() : null;
    }
}
```

But if I instead use this, I don't get any warning about the potential NPE    

```java
@NullUnmarked
public class ResourceConverters {
    public static Resource toResource(MultipartFile file) {
        System.out.println("file = " + file.toString()); // potential NPE
        return file != null ? file.getResource() : null;
    }
}
```

**Comment by agentgt** [comment]
> > BTW If this is new code I highly recommend you do not follow that pattern of nullable in and nullable out especially for public methods.
> 
> @agentgt Presumably you think I should return `Optional` instead of null?
> 

No I'm saying you should not return `null` and take `null` at the same time.

Either do this:

```java
public static @Nullable Resource toResourceOrNull(MultipartFile file)
```

or this:

```java
public static Resource resourceOfNullable(@Nullable MultipartFile file)
```

Ideally this:

```java
public static Resource toResource(MultipartFile file)
```

The idea is to deal with null earlier than later. That is have the caller do this logic:

```java
var resource = file != null ? toResource(file) : null;
```

> @agentgt Based on my testing, using a class-level @NullUnmarked is not the same as annotating each parameter/return value with @Nullable. Specifically, when I use this I correctly get warned by NullAway

My apologies. I did not know you meant the code locally.

I think unspecified is probably up to the tools. For example your code would fail with Eclipse null checker and I think Checkerframework albeit I have not tested with Checkerframework.

**Comment by donalmurtagh** [comment]
>
> The idea is to deal with null earlier than later. That is have the caller do this logic:
> 
> ```java
> var resource = file != null ? toResource(file) : null;
> ```

I don't agree with this. If the method is called 100 times, why burden each caller with checking for null when I could put that logic inside the method (once) instead? To the greatest extend possible, methods should optimise for ease of use, rather than ease of implementation.

**Comment by agentgt** [comment]
> I disagree with this. If the method is called 100 times, why burden each caller with checking for null when I could put that logic inside the method (once) instead? To the greatest extent possible, methods should optimise for ease of use, rather than ease of implementation.

If it works for you and your organization you can just ignore me. However I'll give you this fun exercise... at some point someone has to do something about the `null` right?

> If the method is called 100 times, why burden each caller with checking for null when I could put that logic inside the method (once) instead? 

Well because they are going to have to check the result now afterwards... because the return can be `null`. I'm saying shift that earlier.

> for ease of use

It is not easy. PolyNull is inherently complex topic. Returning an optional or nullable is more complicated for the end user than something that is NonNull.

But again if it works for you then by all means do it and my opinion is probably not germane anyway to the issue.

**Comment by cpovirk** [comment]
For some writing we've done on the idea of a "`@DefaultNullable`" annotation, see these two FAQ entries:

- [Why does being null-marked require us to explicitly mark our nullable types, rather than marking the non-null ones?](https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview?tab=t.0#heading=h.elscixoom7a5)
- [Why not provide more configurable defaults, or at least @DefaultNullable?](https://docs.google.com/document/d/1H3TdqjRZQiGEP7ZePLpTnbYxj9s5hJAn9sp6eOiuj_c/preview?tab=t.0#heading=h.uxjk28dtyjra)

(I've had mixed success with getting my links to take me to the right part of the FAQ, so you may have to scroll.)

**Comment by agentgt** [comment]
@cpovirk was I correct in saying that it is up to the tool to decide or not to fail where @donalmurtagh potential npe comment is?

```java
@NullUnmarked
public class ResourceConverters {
    public static Resource toResource(MultipartFile file) {
        System.out.println("file = " + file.toString()); // potential NPE
        return file != null ? file.getResource() : null;
    }
}
```

That is NullUnmarked is not a SuppressWarnings but it appears nullaway is doing that. I guess what I'm trying to understand is if this is a bug in NullAway and if it is specified somewhere in the spec (ignoring the fun the pun of unspecified).

**Comment by ExE-Boss** [comment]
See also:
- <https://github.com/jspecify/jspecify/issues/79>

**Comment by agentgt** [comment]
I think I had a conversation with @msridhar about something similar:

> agentgt — 07/26/2024 2:10 PM
A useful feature that maybe you guys already have is to detect if the @SuppressWarning("NullAway") is unneeded... eg omit warning say if a flag is passed to the compiler plugin. I think Eclipse does this albeit without a flag.
Basically you run the null analysis on it anyway and if no errors then omit warning. It might even be possible I can try to do a PR for it once I familiarize myself with the project some more.
omit = emit... sorry

> Manu Sridharan — 07/26/2024 2:22 PM
That would definitely be useful.  We rely on Error Prone for handling some suppression annotations like @SuppressWarnings("NullAway") .  Warning on unneeded suppressions of that sort could probably happen at the Error Prone level for any Error Prone check; but we could build some such logic into NullAway also
I think Checker Framework also supports warning on unneeded suppressions


Basically I wanted something similar to what @donalmurtagh is asking for. See Eclipse unlike NullAway and Checkerframework assumes "unspecified" is nullable. 

That is with the exception of Eclipses special annotations you need to not only do `@NullUnmarked` but a `@SuppressWarnings("null")` (I think that is Eclipses suppress string literal) or you need to add the explicit annotations.

I assume such behavior is not against the JSpecify spec?

**Comment by donalmurtagh** [comment]
> 
> If it works for you and your organization you can just ignore me. However I'll give you this fun exercise... at some point someone has to do something about the `null` right?
> 
> > If the method is called 100 times, why burden each caller with checking for null when I could put that logic inside the method (once) instead?
> 
> Well because they are going to have to check the result now afterwards... because the return can be `null`. I'm saying shift that earlier.
> 

@agentgt I don't think these statements are always true. There are some cases where a method returns null and there's no need for the caller to check for nullness in order to prevent a NPE. In the following example assume `User` is a JPA entity class

```java
import java.time.LocalDateTime;

class UserService {

  @Autowired
  private UserRepository userRepository;

  public void updateUser(User user) {
    // there's no need to check if getExpiryTime() returns null
    user.setExpiry(getExpiryTime());
    userRepository.save();
  }

  private @Nuallable LocalDateTime getExpiryTime() {
    var expiry = LocalDateTime.now();
    return expiry.toString().endsWith("0") ? expiry : null;
  }
}
```

**Comment by msridhar** [comment]
Hey, not sure I'm fully following the discussion, but NullAway handles `@NullUnmarked` as follows (this ignores generics, arrays, etc.):

1. `@NullUnmarked` method parameters are treated as `@Nullable` and the return type is treated as `@NonNull`, unless there is an explicit annotation otherwise (like `@NonNull` on a parameter).
2. The implementation of any `@NullUnmarked` method is not checked for nullability errors.

I think the spec only requires 1.  Technically we could not do 2 and require a separate `@SuppressWarnings` annotation, but we did not decide to do that.  Eventually we could re-visit that decision, but I don't think that would solve the original issue outlined here anyway.

**Comment by agentgt** [comment]
> I think the spec only requires 1. Technically we could not do 2 and require a separate @SuppressWarnings annotation, but we did not decide to do that. Eventually we could re-visit that decision, but I don't think that would solve the original issue outlined here anyway.

Yes sorry I roped you in but wanted to make sure my understanding of what NullAway was doing is correct and follows the spec.

While I don't agree with the idea of having `@DefaultNullable` and or the original issue of the pain of having to put `@Nullable` everywhere (because of DRY reasons or whatever) I think there is a migration issue with static checkers that do number 2 in your list.

That is what I think Eclipse is doing is more useful and possible more of what is expected particularly given what @donalmurtagh other complaint is. That is `@NullUnmarked` should not be a `@SuppressWarnings` and I'm concerned that people may come to believe that overtime.

Or perhaps there really needs to be a standard `@SuppressWarnings` if you will.

Otherwise I'm not sure how `@NullUnmarked` is that useful if it is going to vary greatly on behavior of local code. I suppose if you stick to one analyzer you are fine.

> @agentgt I don't think these statements are always true. 

>   private @Nuallable LocalDateTime getExpiryTime() {

That is not polynull like the other examples. I totally expect accessors to potentially have nullables. Anyway don't worry about it. It is more of an internal guideline based on my experience of passing null all over the place which is the case if your API is heavily nullable in and nullable out. An accessor is just nullable out is fine.

There are plenty of reasons to have for example optional in and optional out or nullable in and nullable out but if you are just doing 

```java
@Nullable String blah(@Nullable String x) {
   if (x == null) return null;
   //... 
}
```

It is a code smell for me. If for example `x` gets then set to some default and some search/lookup is done then its ok. Perhaps we can discuss on discord (handle is same as github) as I think I derailed the conversation with that comment/opinion. Apologies.

**Comment by donalmurtagh** [comment]
@msridhar Given that NullAway will warn about a NPE in this case

```java
public class ResourceConverters {
    public static void toResource(@Nullable MultipartFile file) {
        System.out.println("file = " + file.toString()); // NullAway warns about potential NPE
    }
}
```

And NullAway will not warn about a NPE in this case

```java
@NullUnmarked
public class ResourceConverters {
    public static void toResource(MultipartFile file) {
        System.out.println("file = " + file.toString()); // NullAway does NOT warn about potential NPE
    }
}
```

can you explain what you mean by:

> 1. `@NullUnmarked` method parameters are treated as `@Nullable`

given that the 2 examples above are handled differently by NullAway?

**Comment by msridhar** [comment]
> can you explain what you mean by:
> 
> > 1. `@NullUnmarked` method parameters are treated as `@Nullable`
> 
> given that the 2 examples above are handled differently by NullAway?

Sorry I was unclear.  First, `@NullUnmarked` impacts the nullability of the parameter from the perspective of _callers_ to that method.  So for the second example, the nullness of the `file` parameter is "unspecified," which means NullAway allows callers of `toResource` to pass `null`.  Within the body of the method, since `@NullUnmarked` is currently equivalent to suppressing warnings, no checking is done at all; we don't check the body of the method under the assumption that `file` may be `null`.

**Comment by donalmurtagh** [comment]
> since `@NullUnmarked` is currently equivalent to suppressing warnings

@msridhar This is what I missing. Thanks for the explanation

**Comment by cpovirk** [comment]
In an ideal world, it would be possible to run nullness checking on null-unmarked code. (Our reference checker supports this, though it has its own limitations :)) Such a checker could detect the problems with...

```java
String s = null;
// many lines of code that don't give it a real value
return s.length();
```

...or even...

```java
WeakReference<String> s = ...;
return s.get().length();
```

...neither of which requires nullness annotations for the code that is being analyzed. (The latter does require nullness annotations for the JDK's `WeakReference`.) And of course code could be null-unmarked but still have a `@Nullable` annotation on a parameter (as in the examples above), in which case a tool could also identify problems from dereferencing that parameter.

(There is then the _additional_ layer that some tools may choose to report errors for dereferences of _unspecified-nullness_ values, in which case you'd get errors from an unguarded dereference of _any_ parameter in null-unmarked code.)

That said, we [haven't seen it as "wrong"](https://github.com/jspecify/jspecify/issues/236) for a tool to check only null-marked code, and there are obvious ways in which such an approach is desirable.

**Comment by cpovirk** [comment]
(I'll also say that this is another case in which nullness will feel a lot less burdensome if it becomes possible to write "`Foo?`" instead of "`@Nullable Foo`." Hopefully [someday](https://openjdk.org/jeps/8303099)....)

**Comment by agentgt** [comment]
> (I'll also say that this is another case in which nullness will feel a lot less burdensome if it becomes possible to write "Foo?" instead of "@Nullable Foo." Hopefully [someday](https://openjdk.org/jeps/8303099)....)

Likewise most `@PolyNull` methods like `orElse` and `System.getProperty(.., fallback)` could be completely eliminated with the elvis operator (or similar).

(btw if Kevin sees this I can't signup for project Loom's mailinglist for some reason but it would be great if they do not add PolyNull `orElse` to `ScopedValue` and go with `orNull` like Guava optional)

That it a lot of bloat to have all those `orElse` like methods.

**Comment by agentgt** [comment]
> > (btw if Kevin sees this I can't signup for project Loom's mailinglist for some reason but it would be great if they do not add PolyNull `orElse` to `ScopedValue` and go with `orNull` like Guava optional)
> 
> It looks like you got your wish: [openjdk/jdk@4e1878c#diff-1365e5ebb29a505d07d79a37ee24652686046ed1b6d802dc228049a1ef70ce48L615](https://github.com/openjdk/jdk/commit/4e1878ca452d1e781eb61610a94064c0c4798073#diff-1365e5ebb29a505d07d79a37ee24652686046ed1b6d802dc228049a1ef70ce48L615))

Yeah I made a big stink about it on the mailinglist and reddit. However I wanted them to go full nonnull and not even allow `null` in the ScopedValue at all.

So it is a little awkward particularly given `Optional` allows `null` on its `orElse`.

Brian had a sort of complicated but compelling reason why it was best the JDK choose this path. There is probably some Valhalla concerns as well and that will probably make it less awkward. That is presumable they might remove that `requireNonNull` on the `orElse` if we get `T?` I think.

I still think Guava chose the best solution by providing an `orNull()` for `Optional`. I follow similar patterns throughout my OSS code where anything that can return a nullable is suffixed with `orNull` like `somethingOrNull`. I even do it with records. `record Blah(@Nullable Something somethingOrNull) {}`. I even wrote an annotation processor to check the naming scheme.

I have a feeling many do not share the desire to do something that extreme but at least no one can blame me when they get a `null`.

**Comment by kevinb9n** [comment]
This issue has had some wide-ranging discussion; if anyone thinks some subtopics should be hoisted to their own issues please do so!

To the original request: 

If an API has a lot of nullable types in it, then it will have a lot of nullable annotations in it. We believe that is appropriate / working as intended. It may be noisy, but it is non-misleading.

As it is, users need to know how to read two different "kinds" of code: null-marked code and null-unmarked code. We want to keep it to just those two.

I think this issue can be closed?
